pkg/operator/status/condition.go

```diff
@@ -44,8 +44,13 @@ func UnionCondition(conditionType string, defaultConditionStatus operatorv1.Cond
 
 	unionedCondition := operatorv1.OperatorCondition{Type: conditionType, Status: operatorv1.ConditionUnknown}
 	if len(interestingConditions) == 0 {
-		unionedCondition.Status = operatorv1.ConditionUnknown
-		unionedCondition.Reason = "NoData"
+		if conditionType == "Upgradeable" {
```

Comment:
I am thinking of an operator that has a strong opinion on `Upgradeable`. Should we set the default value to `AsExpected` for those operators? (The operator finds an Upgradeable=False condition but fails to update the status for some reason)

Maybe I am missing something here, but to me it feels like that the default `Unknown` is safe to start with here.

Reply:
> The operator finds an `Upgradeable=False` condition but fails to update the status for some reason

I think the flow you're concerned about is:

1. API server is unavailable, operator fails to write some `UpgradeableWhatever=False` operator condition.
2. API server becomes available, operator forgets that it just wanted to write `UpgradeableWhatever=False`, doesn't see anything concerning on the in-cluster status, and writes `Upgradeable=True` `AsExpected`.
3. API server remains available, operator successfully writes `UpgradeableWhatever=False`.
4. API server remains available, operator successfully notices `UpgradeableWhatever=False` and writes `Upgradeable=False`.

I'm a bit fiddly on how client-side caching works.  Ideally "forgets that it just wanted to write" would not be possible.  If that's possible today, library-go should be able to add additional client-side guards to work around it.  But even if it is possible today, it seems unlikely to me that the gap between 1 and 4 would be all that large or likely to receive and update request.  And if that gap exists, you're still exposed to the gap between 1 and 3 in:

1. API server is available, operator writes `UpgradeableWhatever=False`.
2. API server becomes unavailable, operator cannot write `Upgradeable=False`.
3. API server becomes available, operator successfully notices `UpgradeableWhatever=False` and writes `Upgradeable=False`.

So I'm not sure how much effort we want to address in driving down this latency.  It's hard to imagine cases that would be really terrible for an update that did not exist minutes before the update was requested.  And if those sorts of cases do exist, they could also occur minutes after the update was requested when we were already committed.  Thoughts?

Reply:
My apology, I misunderstood the changes. So if I am not mistaken, `UnionCondition` aggregates the `*Upgradeable` conditions reported by an operator and it synthesizes a new `Upgradeable` condition to be written into `ClusterOperator` status. 
And by default, if `len(interestingConditions)` is zero then we set `Upgradeable` to `Unknown`. 

So if we have an operator that does not have *any* opinion on `Upgradeable`:
`interestingConditions` will be empty -> `Upgradeable` is set to `Unknown` -> blocks upgrade?

If I have understood this correctly, let me know if the following scenario makes sense:
- We have an operator that has a strong opinion on `Upgradeable` and it maintains only one condition type for that -  `Foo_Upgradeable` condition. Due to a bug in the code it stops writing this condition. When we aggregate inside `UnionCondition` function, `interestingConditions` will be empty and we will write `AsExpected`.  (our defaulting was optimistic and led to this issue)  

A couple of points:
- In general, I think we should always set the default value to be pessimistic or `Unknown` in this case so we don't let a bug go unnoticed. we don't automatically conclude that it's working as expected, when it may not be the case.
- Not have special handling of a certain condition type inside `UnionCondition`. This change is not desirable for an operator that has a strong opinion of `Upgradeable`, right? Maybe we can parameterize this change so each author can choose the default value appropriately or add a new function that writes `AsExpected`. So it becomes more of an opt in process vs automatic.

Thoughts?

Reply:
> So if we have an operator that does not have any opinion on Upgradeable: interestingConditions will be empty -> Upgradeable is set to Unknown -> blocks upgrade?

That's all right except "blocks updates".  The CVO only blocks minor-version updates on `Upgradeable=False`.  The issue with `Upgradeable=Unknown` is just that [folks find it confusing][1].

> We have an operator that has a strong opinion on `Upgradeable` and it maintains only one condition type for that - `Foo_Upgradeable` condition. Due to a bug in the code it stops writing this condition.

There's no freshness cutoff in any of these conditions.  If you stop writing the condition, this union code will continue to pass along whatever value you'd set before when it updates `Upgradeable`.

> When we aggregate inside `UnionCondition` function, `interestingConditions` will be empty and we will write `AsExpected`.

It will only be empty if the operator has not written any `*Upgradeable` operator conditions for the union code to consume.  That's possible, but I don't think guarding against it is worth setting `Upgradeable=False`, and without doing that, you get no protection from the CVO.

> This change is not desirable for an operator that has a strong opinion of `Upgradeable`, right?

This change is irrelevant for operators with opinions, because they will always have content in `interestingConditions`.  The way you express "I have an opinion" is just by setting `*Upgradeable` operator conditions.  I don't think we need a parallel knob to emphasize that.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1928141#c0